File: Backend/models/Grade.js
const Sequelize = require('sequelize');
const sequelize = require('../config/database');
const Student = require('./Student');
const Course = require('./Course');
const Semester = require('./Semester');
const GPARule = require('./GPARule');

const Grade = sequelize.define('Grades', {
  grade_id: {
    type: Sequelize.INTEGER,
    autoIncrement: true,
    primaryKey: true,
  },
  student_id: {
    type: Sequelize.STRING(10),
    allowNull: false,
    references: {
      model: Student,
      key: 'student_id',
    },
  },
  course_id: {
    type: Sequelize.STRING(10),
    allowNull: false,
    references: {
      model: Course,
      key: 'course_id',
    },
  },
  semester_id: {
    type: Sequelize.INTEGER,
    allowNull: false,
    references: {
      model: Semester,
      key: 'semester_id',
    },
  },
  grade: {
    type: Sequelize.STRING(2),
    defaultValue: null,
  },
  percentage: {
    type: Sequelize.INTEGER,
    defaultValue: null,
  },
  course_grade_points: {
    type: Sequelize.DECIMAL(4, 1),
    defaultValue: null,
  },
}, {
  timestamps: false,
  tableName: 'Grades',
  hooks: {
    beforeCreate: async (grade, options) => {
      if (grade.grade) {
        const gpaRule = await GPARule.findOne({ where: { letter_grade: grade.grade } });
        if (!gpaRule) throw new Error('Invalid grade: No matching GPA rule found');

        const course = await Course.findByPk(grade.course_id);
        if (!course) throw new Error('Course not found');

        grade.course_grade_points = (gpaRule.gpa_points * course.credits).toFixed(1);
      }
    },
    beforeUpdate: async (grade, options) => {
      if (grade.changed('grade')) {
        const gpaRule = await GPARule.findOne({ where: { letter_grade: grade.grade } });
        if (!gpaRule) throw new Error('Invalid grade: No matching GPA rule found');

        const course = await Course.findByPk(grade.course_id);
        if (!course) throw new Error('Course not found');

        grade.course_grade_points = (gpaRule.gpa_points * course.credits).toFixed(1);
      }
    },
  },
});

module.exports = Grade;

File: grade-simulator-service/dummy_data/models/grade.py
from sqlalchemy import Column, String, Integer, ForeignKey, Numeric
from sqlalchemy.orm import relationship
from db import Base
from decimal import Decimal

class Grade(Base):
    __tablename__ = "Grades"
    
    grade_id = Column(Integer, primary_key=True, autoincrement=True)
    student_id = Column(String(10), ForeignKey("Students.student_id"), nullable=False)
    course_id = Column(String(10), ForeignKey("Courses.course_id"), nullable=False)
    semester_id = Column(Integer, ForeignKey("Semesters.semester_id"), nullable=False)
    grade = Column(String(2), nullable=True)
    percentage = Column(Integer, nullable=True)
    course_grade_points = Column(Numeric(4, 1), nullable=True)
    
    def to_dict(self):
        return {
            "grade_id": self.grade_id,
            "student_id": self.student_id,
            "course_id": self.course_id,
            "semester_id": self.semester_id,
            "grade": self.grade,
            "percentage": self.percentage,
            "course_grade_points": float(self.course_grade_points) if isinstance(self.course_grade_points, Decimal) else self.course_grade_points
        } 